### PR TITLE
Refactor `absolute_paths`

### PR DESCRIPTION
**Pull request by Jarcho**
Checks are rearranged to do the more expensive checks later. Since the most likely path length will be one (locals and imported/local items) this will exclude such paths on the first check.

Tests were rewritten as they were hard to follow (annotations would have helped), spammy (lots of tests for the same thing) and insufficient.

One thing thing that came up and should be decided on now is what to do about the difference between `path::to::Trait::item` (4 segments) and `path::to::Type::item` (3 segments). The current behaviour treats these as different lengths which is terrible. I personally think these should both be three segments since the item can't actually be imported. Only the type or the trait could be. This makes `crate_name::Trait::item` the shortest absolute path which is shorter than the lint allows by default.

changelog: None

### CODE REVIEW

File: tests/ui-toml/absolute_paths/absolute_paths.rs [hunk 1]
```diff
@@ -1,97 +1,120 @@
...
-    helper::a();
-    use crate::a::b::c::C;
-    use a::b;
-    use std::f32::MAX;
-    a::b::c::d::e::f::F;
-    b::c::C;
-    fn a() -> a::A {
-        todo!()
+    // FIXME: `path` in `<_ as path>` doesn't actually exist in the HIR tree
+    let _ = <i32 as core::clone::Clone>::clone(&0i32);
```
Comment (y21) [inline tests/ui-toml/absolute_paths/absolute_paths.rs:43]: Looking at the hir-tree, it does seem to exist: https://godbolt.org/z/h66erEjq3
There's a path in there with the segments `std::clone::Clone::clone`

File: clippy_lints/src/absolute_paths.rs [hunk 2]
```diff
@@ -47,54 +56,40 @@ impl_lint_pass!(AbsolutePaths => [ABSOLUTE_PATHS]);
...
-            && let (segment, len) = if first.ident.name == kw::PathRoot {
-                // Indexing is fine as `PathRoot` must be followed by another segment. `len() - 1`
-                // is fine here for the same reason
-                (&rest[0], path.segments.len() - 1)
-            } else {
-                (first, path.segments.len())
-            }
-            && len > *absolute_paths_max_segments as usize
-            && let Some(segment_snippet) = snippet_opt(cx, segment.ident.span)
-            && segment_snippet == segment.ident.as_str()
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_lints/src/absolute_paths.rs:78]: That's the only thing I can think of. First version of the lint (never merged) used `is_from_proc_macro` and didn't have a check like this.

File: clippy_lints/src/absolute_paths.rs [hunk 3]
```diff
@@ -47,54 +56,40 @@ impl_lint_pass!(AbsolutePaths => [ABSOLUTE_PATHS]);
...
-    fn check_path(&mut self, cx: &LateContext<'_>, path: &Path<'_>, hir_id: HirId) {
-        let Self {
-            absolute_paths_max_segments,
-            absolute_paths_allowed_crates,
-        } = self;
+    fn check_path(&mut self, cx: &LateContext<'tcx>, path: &Path<'tcx>, hir_id: HirId) {
+        let (first, len) = match path.segments {
+            // Any path starting with `::` must have the second segment refer to an external crate.
+            // Make sure not to include the `::` segment when counting the segments.
+            [root, first, ..] if root.ident.name == kw::PathRoot => (first, path.segments.len() - 1),
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_lints/src/absolute_paths.rs:70]: Right. I've blocked that insanity out of my memory.

File: tests/ui-toml/absolute_paths/absolute_paths.rs (hunk 1 above)
Comment (Jarcho) [reply tests/ui-toml/absolute_paths/absolute_paths.rs:43]: I had assumed it was being picked up as a type relative path. The issue seems to be the span of path covering the entire `QPath` which causes the proc macro check to fail.

**Comment by Jarcho** [comment]
Went ahead and changed `path::to::Enum::variant` and `path::to::Trait::item` to both not count the final element (both would be counted as three, not four).

Also make sure to ping me to squash this before merging. Those last two commit messages are useless.

**Review by y21 — APPROVED** [review]
Took me a fair bit to go through the test changes but looks good (aside from 2 minor things). What you said with not counting the last segment for type/trait associated items also makes sense to me.

File: clippy_utils/src/check_proc_macro.rs [hunk 4]
```diff
@@ -123,16 +123,14 @@ fn qpath_search_pat(path: &QPath<'_>) -> (Pat, Pat) {
 
 fn path_search_pat(path: &Path<'_>) -> (Pat, Pat) {
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_utils/src/check_proc_macro.rs:124]: Thankfully we can't encounter those in most places since they can only be generic constraints. I'll be rewriting the proc macro detection anyways since the current approach isn't great.

File: tests/ui-toml/absolute_paths/absolute_paths.rs [hunk 5]
```diff
@@ -1,97 +1,122 @@
...
+    // Treated as having three parts.
+    let _ = ::core::clone::Clone::clone(&0i32);
+    //~[default]^ absolute_paths
+    //~[no_short]| absolute_paths
+    let _ = <i32 as core::clone::Clone>::clone(&0i32);
+    //~[default]^ absolute_paths
+    //~[no_short]| absolute_paths
+    let _ = std::option::Option::None::<i32>;
+    //~[default]^ absolute_paths
+    //~[no_short]| absolute_paths
```
Comment (y21) [inline tests/ui-toml/absolute_paths/absolute_paths.rs:45]:
```suggestion
    //~[no_short]| absolute_paths
    //~[allow_crates]| absolute_paths
```

**Review by y21 — APPROVED** [review]
There's one missing test annotation that I missed last time (normally I wouldn't comment on this since I don't *think* we really require them yet, but since you already went out of your way to annotate it all). All other changes look good to me.

r=me either way
